contracts/Oracle/Calculations/Balancer.sol

```diff
@@ -0,0 +1,135 @@
+    function getPriceUsdcRecommended(address tokenAddress)
+        external
+        view
+        returns (uint256);
+    function usdcAddress() external view returns (address);
+}
+
+interface IYearnAddressesProvider {
+    function addressById(string memory) external view returns (address);
+}
+
+contract CalculationsBalancer is Ownable {
+    IYearnAddressesProvider internal yearnAddressesProvider;
+
+    IBalancerVault internal constant balancerVault =
+        IBalancerVault(0xBA12222222228d8Ba445958a75a0704d566BF2C8);
+
+    constructor(address _yearnAddressesProviderAddress) {
+        yearnAddressesProvider = IYearnAddressesProvider(
+            _yearnAddressesProviderAddress
+        );
+    }
+
+    function updateYearnAddressesProviderAddress(
+        address _yearnAddressesProviderAddress
+    ) external onlyOwner {
+        yearnAddressesProvider = IYearnAddressesProvider(
+            _yearnAddressesProviderAddress
+        );
+    }
+
+    function oracle() internal view returns (IOracle) {
+        return IOracle(yearnAddressesProvider.addressById("ORACLE"));
+    }
+
+    //Phantom pools
+    //Stable Pools
+    function poolLpTotalValueUsdc(address lpAddress)
+        public
+        view
+        returns (uint256)
+    {
+        //If the token is a Wrapped Aave asset return the total atoken value held
+        try IWrappedAsset(lpAddress).ASSET() returns (address _asset) {
+            IERC20 aToken = IERC20(IWrappedAsset(lpAddress).ATOKEN());
+            return
+                (oracle().getPriceUsdcRecommended(_asset) *
+                    aToken.balanceOf(lpAddress)) / (10**aToken.decimals());
+        } catch {}
+
+        bytes32 poolId = IBalancerPool(lpAddress).getPoolId();
+
+        (IERC20[] memory tokens, uint256[] memory balances, ) = balancerVault
+            .getPoolTokens(poolId);
+        uint256 totalValue;
+        for (uint256 i; i < tokens.length; i++) {
+            address token = address(tokens[i]);
+
+            if(token == lpAddress) continue;
+
+            uint256 tokenValueUsdc = getTokenAmountUsdc(token, balances[i]);
+            totalValue += tokenValueUsdc;
+        }
+        return totalValue;
+    }
+
+    function getTokenAmountUsdc(address _token, uint256 _amount)
+        public
+        view
+        returns (uint256)
+    {
+        uint8 decimals = IERC20(_token).decimals();
+        uint256 tokenPrice = oracle().getPriceUsdcRecommended(_token);
+        uint256 tokenValueUsdc = (_amount * tokenPrice) / 10**decimals;
+        return tokenValueUsdc;
+    }
+
+    function poolLpPriceUsdc(address lpAddress) public view returns (uint256) {
+        uint256 totalValueUsdc = poolLpTotalValueUsdc(lpAddress);
+        
+        IBalancerPool pool = IBalancerPool(lpAddress);
+
+        uint256 totalSupply = pool.totalSupply();
+        //If its a phantom pool we need to use "actual supply" or virtual supply
+        try pool.getVirtualSupply() returns (uint256 _supply) {
```

Comment:
Is there a way to differentiate these two types of pools without try-catch? would be nice if this could just be an `if-else`

Reply:
Not that I can figure out at least without making it way more complicated. There are alot of different types of pools where even the phantom pools need different getters for supply. Other than reading strings from the symbols/names, which seems sketchy since anyone can deploy a balancer pool and there is no real standard, or doing all of the fee math manually for every pool I dont see a better way.